Mocked SSH server and enable ssh test on windows

ag613915cao opened the pull request:
- Keep current SSH remote connection test on real system
- Faked and Mocked SSH server which running on both Linux and Windows
- Add corresponding tests

Fixed #1704

efiop commented on tests/func/test_data_cloud.py:
```diff
@@ -101,11 +101,11 @@ def _should_test_ssh():
-
     try:
+        if os.name == "nt":
+            check_output(
```
Why are we running it twice? It is still raising the error, so should_test returns False. This is actually one of very early tests, that has a drawback of not raising `SkipTest` exception and instead relying on `if should_test`. You could see that ssh tests are still not run by looking at coverage reports on appveyor and travis. On travis:
```
dvc/remote/ssh/__init__.py        81      7    91%
dvc/remote/ssh/connection.py     175     21    88%
```
on appveyor
```
dvc\remote\ssh\__init__.py        81     44    46%
dvc\remote\ssh\connection.py     175     99    43%
```

efiop replied:
Btw, if your patch is not yet ready for review, please feel free to use `[WIP]` prefix. :slightly_smiling_face:

ag613915cao replied:
@efiop  Yes. I know about that. Still in progress of investigating

ag613915cao commented:
@efiop I have just supported Mocked SSH server which is not rely on OS ssh binary. And the test case on Linux is passed. 
Will to check with windows OS case soon.
Meanwhile, could you please check if it is the right way? Thanks

ag613915cao commented:
@efiop Do you know why it is timeout for `walk_files` while running the test on Windows [1]? I have check the NOTE but not quite understand [2]

[1] https://ci.appveyor.com/project/iterative/dvc/builds/24039752#L2917
[2] https://github.com/iterative/dvc/blob/master/dvc/remote/base.py#L295-L299

efiop commented:
@ag613915cao I believe that timeout is not related to that note in our code. You should probably simply try increasing the timeout in `tests/__main__.py`. At least that would be the first thing that I would try :slightly_smiling_face:

ag613915cao commented:
> 
> 
> @ag613915cao I believe that timeout is not related to that note in our code. You should probably simply try increasing the timeout in `tests/__main__.py`. At least that would be the first thing that I would try 🙂

Thanks @efiop . I tried to increased timeout from 600 to 1800 but it is still failed. Let's me try to increase it more to see.
For the source, I tried to debug and it hang at `paramiko.buffered_pipe` in `read` method. Maybe the file is too big with the timeout.

efiop commented:
@ag613915cao got it. It was simply my first instinct :slightly_smiling_face: Can't say what happened here right away, need to take a closer look :slightly_frowning_face:

efiop commented:
@ag613915cao Timeout of 9000 is way too much :) I was wondering if maybe increasing it a bit would help, but waiting for several hours to run this test is not going to tell us anything. Btw, please rebase on top of master, there were a few changes related to `walk` method, maybe they will help.

ag613915cao commented:
@efiop Indeed! The reason I set it to 9000 to see if it is move forward. Because I couldn't debug it in local without any hints or source of reason. 
I have just rebased on top of latest master branch (I just rebased it couple hours ago and didn't aware of new landing patch related to `walk` method) to see if it help (and timeout is set to 3600 for reasonable).

ag613915cao commented:
@efiop After prepared local environment I found the root cause. It is not related to our code base as you mentioned. It is just because `prefix` in Windows contain too much dirs and files for processing (especially `C:\Windows` directory).
I have updated the code (no need to extend timeout) and it is ready to go. 
So, kindly please take a look. Thank you

efiop commented on dvc/remote/ssh/__init__.py:
```diff
@@ -43,7 +46,12 @@ def __init__(self, repo, config):
             or getpass.getuser()
         )
-        self.prefix = parsed.path or "/"
+        # In Windows, there are too much dirs and file will be fetch
```
So which test failed? I don't think data_cloud test tries to walk root. If it does then it is a bug. And we shouldn't use such hack in the RemoteSSH itself.

ag613915cao replied:
You are right. This is a bug from urlparse in windows. It always return '' for `path`. Then it leaded to root directory. In Windows root directory (assume it is `C:` path), there are some directories are protected by the OS.
But please disregard this point as I fixed the urlparse on Windows in next patch already.

efiop commented on dvc/remote/ssh/connection.py:
```diff
@@ -135,24 +135,28 @@ def walk(self, directory, topdown=True):
-                dirs.append(name)
+            # In Windows some directories are protected by the OS. Skipped!
+            if os.name == "nt":
+                if (exc.errno == 13) or (exc.errno is None):
```
Please don't hardcode errnos. Use errno module instead.

efiop commented on dvc/remote/ssh/connection.py:
```diff
@@ -135,24 +135,28 @@ def walk(self, directory, topdown=True):
-            if S_ISDIR(entry.st_mode):
-                dirs.append(name)
+            # In Windows some directories are protected by the OS. Skipped!
+            if os.name == "nt":
```
This is a pretty strange check. So what exactly raises that error? listdir_attr? If so, why wrap the whole hoop in this try-except? Also, i don't think we should ignore EPERM like this, since it would be an unexpected behavior for the user. If he has supplied a remote with protected directories we should raise an error and let him know about it, and not hide it away.

ag613915cao replied:
Yes. The above answered applied to this concern too. I dropped this in next patch

efiop commented on tests/func/test_data_cloud.py:
```diff
@@ -112,6 +115,21 @@ def _should_test_ssh():
+        return True
+
+    try:
+        if os.name == "nt":
```
why have two different commands to test on windows and *nix? Why not use one? Also, the point of this PR is to make ssh test everywhere and always. So it doesn't need this function since it always should return True.

ag613915cao replied:
How do you run `ls` command in Windows? AFAIK, the command is not supported.

efiop replied:
But `dir` works on linux :) At least that is how it is usually.

ag613915cao replied:
Cool, I understand your point now ;)

efiop commented on tests/func/test_data_cloud.py:
```diff
@@ -149,6 +167,18 @@ def get_ssh_url():
+def get_ssh_url_mocked(user, port):
+    if os.name == "nt":
+        url = "ssh://{}@127.0.0.1:{}:{}".format(
+            user, port, get_local_storagepath(directory=here)
```
I don't think it should create dirs in `here`. Why `get_local_storagepath` doesn't work?

ag613915cao replied:
Cool, you are right.

efiop commented on tests/func/test_data_cloud.py:
```diff
@@ -395,6 +430,50 @@ def _get_cloud_class(self):
+
+
+@pytest.yield_fixture()
+def server():
```
we already have this (or actually `ssh()` fixture that I've introduced instead) fixture in `tests/unit/remote/ssh`, no need to copypaste it.

ag613915cao replied:
Cool. I moved the `ssh()` to conftest.py

efiop commented on tests/func/test_data_cloud.py:
```diff
@@ -395,6 +430,50 @@ def _get_cloud_class(self):
+
+    @staticmethod
+    def _client(server):
+        client = SSHConnection(
```
Again, please use `ssh` fixture from tests/unit/remote/ssh .

ag613915cao replied:
DONE

efiop commented on tests/func/user.key:
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
You already added it in `tests/unit/remote/ssh`, why copy paste it? Just put it somewhere where both tests can find it and use it.

ag613915cao replied:
DONE

efiop commented on tests/func/user.key.pub:
```diff
@@ -0,0 +1 @@
+ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQD8eUxHRyGAUvHoD8GZiL1imP2evSRndajqfdl1Jc5Elofp7t0tLD7R6gPU0MLEM5FmXBg0CDjnOd1Tmw83gi9VwC8L8UPSpIrUfSS7WnhBXzbLhrVI6fNGUNQhWF1DGeVBryAj7MM7wDLM4nihn0svsfL6YgOobFHAd4p8Nhe4eQT5oz0IKSJ0+pH9SrFqDTf7LnUwSmELTuavuasNviU6Z6rLxerRlxSMiSYqIYJVVEk7qDiRgXCvRTCkoLAvM5/DgXLZKnNZ97ch1VSoiuj3t3UAYL705RAROJWkFg5sEjWp9Ch5/oxP4IVxNXjRzIp0IbrHnt/qCk8xiph90dPJ user@unicorn
```
Same as with user.key.

ag613915cao replied:
DONE

ag613915cao commented:
@efiop I have updated the patch based on your comments. Kindly please take a look and let me know if it need to update something. Thanks a lot

ag613915cao commented:
Just resolved merge conflict (no logic change)

efiop commented on dvc/remote/ssh/__init__.py:
```diff
@@ -43,7 +44,7 @@ def __init__(self, repo, config):
             or getpass.getuser()
         )
-        self.prefix = parsed.path or "/"
+        self.prefix = url2pathname(parsed.path) or "/"
```
Hm, but if you pass ntpath here, then everything will break down the line since we use posixpath to compose paths. So with `self.prefix = C:\some\path` we'll get something like `C:\some\path\/new/path`, which is not right.

The problem comes from the fact that we didn't actually ever think about connecting _to_ windows server and using ntpaths there. The scenario always was that you connect from windows(or any other client) to *nix host that uses normal posixpaths. I'm wondering if we should simply forbid ntpaths in ssh urls. Does mockssh(or, underlying paramiko server) support posixpaths when running on windows? If so, is there a way to create a temp dir and address it with posixpath in our tests?

ag613915cao replied:
I see.

>Does mockssh(or, underlying paramiko server) support posixpaths when running on windows?

I think mockssh is designed to run not against any OS. So, maybe it is not support posixpaths when running on windows

>If so, is there a way to create a temp dir and address it with posixpath in our tests?

Let me check if we can address the temp dir with posixpath from `get_local_storagepath()`.

Do you have any suggestion to solve this problem?

efiop replied:
Windows actually supports `/` in paths, see https://en.wikipedia.org/wiki/Path_%28computing%29#MS-DOS.2FMicrosoft_Windows_style . So maybe something like `get_local_storagepath().replace('\\', '/')` would work.

ag613915cao replied:
Cool. It solved the problem :)

ag613915cao commented:
@efiop I updated the patch with your suggestion :) Kindly please help to take a look and land it if possible.

efiop commented on tests/func/test_data_cloud.py:
```diff
@@ -149,6 +152,18 @@ def get_ssh_url():
 
+def get_ssh_url_mocked(user, port):
+    if os.name == "nt":
+        url = "ssh://{}@127.0.0.1:{}:{}".format(
```
I don't think this second `:` is necessary. Should it be something like
```
path = get_local_storagepath()
if os.name == "nt":
    path = "/" + path.replace("\\", "/")
url = "ssh://{}@127.0.0.1:{}{}".format(user, port, path)
``` 
? Also please don't forget to add a note here on why you are doing that `replace()` magic.

ag613915cao replied:
No, It will raise an exception: ` DvcException: couldn't get the '/c:/users/cao/appdata/local/temp/dvc-test.7968.outdjx.e8373c98-b117-479c-9ed3-000a974182da' remote directory files list` . Because the initial slash `/` before `c:/`. That is the reason I provide this indicator to solve in Windows. Is that make sense or do you have other solution?

ag613915cao replied:
> Also please don't forget to add a note here on why you are doing that replace() magic.

I will append the note

efiop commented on tests/unit/remote/ssh/test_ssh.py:
```diff
@@ -30,15 +30,17 @@ def test_url(self):
 
         # URL ssh://[user@]host.xz[:port]/path
-        url = "ssh://{}@{}:{}{}".format(user, host, port, path)
+        if os.name == "nt":
```
is this still necessary?

ag613915cao replied:
Yes. It is necessary as I answered in above comment. In windows, we need an indicator to separate between `port` and `path`. If not the `port` will get `c` drive character and the make the port validation failed as it contain a character in it.

efiop replied:
And now this is no longer needed, right?

ag613915cao replied:
Oh, You read my mind. I was pushed new update while you posted the above comment :)
DONE

efiop commented on tests/func/test_data_cloud.py:
```diff
@@ -229,9 +247,11 @@ def _setup_cloud(self):
 
         config = copy.deepcopy(TEST_CONFIG)
         config[TEST_SECTION][Config.SECTION_REMOTE_URL] = repo
+        config[TEST_SECTION][Config.SECTION_REMOTE_KEY_FILE] = keyfile
```
Btw, you are mocking the connection itself, so keyfile, port, user, etc options passed through here don't even have any effect. Should we only start the ssh server and not mock the connection itself, so that RemoteSSH establishes the connection with the server based on these options by itself?

ag613915cao replied:
Indeed. Actually, this was my initial proposal to not mock connection itself and left server open for further adjust. 
I will update this point.

ag613915cao replied:
DONE

ag613915cao commented:
@efiop I updated the patch as you requested. Kindly please take a look and let me know if it needs to update something. Thanks a lot :)

efiop commented on tests/conftest.py:
```diff
@@ -57,3 +61,25 @@ def dvc(repo_dir, git):
+
+
+@pytest.yield_fixture()
+def ssh():
```
You could use `ssh_server` fixture in this fixture, so you don't duplicate code :)

ag613915cao replied:
Cool. DONE

efiop commented on tests/func/test_data_cloud.py:
```diff
@@ -149,6 +152,21 @@ def get_ssh_url():
 
+def get_ssh_url_mocked(user, port):
+    if os.name == "nt":
+        # since we use posixpath to compose paths.
```
I believe this comment could be improved :slightly_smiling_face:

ag613915cao replied:
What do you want to improve? I'm not a native speaker, so description sometimes not ideal for me.

efiop commented on tests/func/test_data_cloud.py:
```diff
@@ -149,6 +152,21 @@ def get_ssh_url():
+        # like C:\some\path\/new/path, which is not right.
+        # So, we need to make the path consistence
+        path = get_local_storagepath().replace("\\", "/")
+        url = "ssh://{}@127.0.0.1:{}:{}".format(user, port, path)
```
Continuing from https://github.com/iterative/dvc/pull/1911#discussion_r278794297 .

> No, It will raise an exception: DvcException: couldn't get the '/c:/users/cao/appdata/local/temp/dvc-test.7968.outdjx.e8373c98-b117-479c-9ed3-000a974182da' remote directory files list . Because the initial slash / before c:/. That is the reason I provide this indicator to solve in Windows. Is that make sense or do you have other solution?

But you are not passing the leading slash at all right now. What you are currenty using as an url is `ssh://example.com:88:c:/path/to`, which when parsed by urlparse results in:
```
>>> urlparse.urlparse("ssh://example.com:88:c:/path/to")
ParseResult(scheme='ssh', netloc='example.com:88:c:', path='/path/to', params='', query='', fragment='')
```
so your prefix is `/path/to` and not `c:/path/to`. I'm wondering if it is even accessing the proper directory.

And if you remove `:` as I'm asking, you'll be passing `ssh://example.com:88c:/path/to`, which will result in
```
>>> urlparse.urlparse("ssh://example.com:88c:/path/to")
ParseResult(scheme='ssh', netloc='example.com:88c:', path='/path/to', params='', query='', fragment='')
```
which is totally wrong.

efiop replied:
The reason why it worked is probably because paramiko server is running in `C:\`, so it becomes its root and `/path/to` is resolved relative to it. Which makes sense and is indeed in sync with how ssh servers work on windows. So we should use not `c:/path/to`, but rather `/path/to` in our urls(e.g. `ssh://example.com:22:/path/to` or `ssh://example.com:22/path/to`).

ag613915cao replied:
Hm. I don't quite understand what you want me to change about this point. Could you elarboate more? Thank you

ag613915cao replied:
Ah, I got your point now

efiop replied:
Your ssh server has root set to `C:\\`, so it will resolve paths relative to that. Which means that you need to use `/path/to` as `path` in the `url`.

ag613915cao replied:
DONE, sir :)

ag613915cao commented:
@efiop I think all the points has been addressed :)

efiop commented on tests/func/test_data_cloud.py:
```diff
@@ -149,6 +152,17 @@ def get_ssh_url():
+        # Since we use posixpath to compose paths.
+        # In Windows the path would look like C:\\some\\path
+        # We need to make the path consistence
+        path = path.split(":")[1].replace("\\", "/")
```
How about
```suggestion
        # NOTE: On Windows get_local_storagepath() will return an ntpath
        # that looks something like `С:\some\path`, which is not compatible
        # with SFTP paths [1], so we need to convert it to a proper posixpath.
        # To do that, we should construct a posixpath that would be relative
        # to the server's root. In our case our ssh server is running with `C:/`
        # as a root, and our URL format requires absolute paths, so the
        # resulting path would look like `/some/path`.
        #
        # our SSH remote URLs require absolute paths, so w
        # [1] https://tools.ietf.org/html/draft-ietf-secsh-filexfer-13#section-6
        drive, path = os.path.splitdrive(path)
        assert drive == "C:/"
        path = path.replace("\\", "/")
```

ag613915cao replied:
Yes. I updated as your expectation

efiop commented on tests/unit/remote/ssh/test_ssh.py:
```diff
@@ -38,7 +37,6 @@ def test_url(self):
         self.assertEqual(remote.user, user)
         self.assertEqual(remote.host, host)
         self.assertEqual(remote.port, port)
-        self.assertEqual(remote.prefix, path)
```
Is this needed?

ag613915cao replied:
My bad. It was wrong when I rebased with old version of code

ag613915cao replied:
DONE

efiop commented on tests/unit/remote/ssh/test_ssh.py:
```diff
@@ -1,5 +1,4 @@
 import getpass
-
```
Let's not remove this line, since it separates test modules from getpass.

ag613915cao replied:
DONE.

ag613915cao commented:
@efiop  I'm happy to work with you. :)

Suor commented:
@ag613915cao `@pytest.yield_fixture` is deprecated now, you use normal `@pytest.fixture`, it sees `yield` and does the right job.

ag613915cao commented:
@Suor I don't know who you are and don't like this working style. If you feel it needs to improve, please open new issue. Or at least, please provide feedback before the patch is merged.
You just gave your comment without any evidence/reference. That is not good in my opinion. Thanks!

shcheklein commented on tests/conftest.py:
```diff
@@ -57,3 +61,26 @@ def dvc(repo_dir, git):
+key_path = os.path.join(here, "{0}.key".format(user))
+
+
+@pytest.yield_fixture()
```
Hi @ag613915cao , I think @Suor was referring to using this decorator instead of just `@pytest.fixture`. See the example in this file, a few lines above.

ag613915cao replied:
@shcheklein Yes. I know about that. What I mean here is the way reviewer provide feedback to contributor(s). I have at least 5 years contributing to open source projects. I can realize that.

For this point: I will update this within today. Thank you for your consideration.

Suor commented:
@ag613915cao I didn't mean to offend you. I saw a deprecated decorator, [rechecked it](https://docs.pytest.org/en/latest/yieldfixture.html) and posted the comment here for this to not get lost.

I understand that it would be easier if I'd posted this before PR got merged, but I only saw it after.

ag613915cao commented:
@Suor Sure. I will update the point within today to make you feel satisfaction. Thank you.

`Marking functions as yield_fixture is still supported, but deprecated and should not be used in new code.`

[1] https://docs.pytest.org/en/latest/yieldfixture.html
